File: 2015/Day14/main.py
import re

def parse_reindeer(line):
    pattern = r'(\w+) can fly (\d+) km/s for (\d+) seconds, but then must rest for (\d+) seconds.'
    match = re.match(pattern, line)
    if match:
        name, speed, fly_time, rest_time = match.groups()
        return name, int(speed), int(fly_time), int(rest_time)
    return None

def calculate_distance(speed, fly_time, rest_time, total_time):
    cycle_time = fly_time + rest_time
    full_cycles = total_time // cycle_time
    remaining_time = total_time % cycle_time

    distance = full_cycles * fly_time * speed
    if remaining_time > fly_time:
        distance += fly_time * speed
    else:
        distance += remaining_time * speed

    return distance

def calculate_distance2(speed, fly_time, rest_time, current_time):
    cycle_time = fly_time + rest_time
    full_cycles = current_time // cycle_time
    remaining_time = current_time % cycle_time

    distance = full_cycles * fly_time * speed
    if remaining_time > fly_time:
        distance += fly_time * speed
    else:
        distance += remaining_time * speed

    return distance

def find_winning_reindeer(reindeer_list, total_time):
    max_distance = 0
    winner = None

    for reindeer in reindeer_list:
        name, speed, fly_time, rest_time = reindeer
        distance = calculate_distance(speed, fly_time, rest_time, total_time)
        if distance > max_distance:
            max_distance = distance
            winner = name

    return winner, max_distance

def simulate_race(reindeer_list, duration):
    points = {reindeer[0]: 0 for reindeer in reindeer_list}
    distances = {reindeer[0]: 0 for reindeer in reindeer_list}

    for second in range(1, duration + 1):
        for reindeer in reindeer_list:
            name, speed, fly_time, rest_time = reindeer
            if second % (fly_time + rest_time) <= fly_time and second % (fly_time + rest_time) != 0:
                distances[name] += speed

        max_distance = max(distances.values())
        for reindeer in reindeer_list:
            name = reindeer[0]
            if distances[name] == max_distance:
                points[name] += 1

    return points

with open('input.txt', 'r') as file:
    reindeer_data = file.readlines()

reindeer_list = [parse_reindeer(line) for line in reindeer_data]
winner, distance = find_winning_reindeer(reindeer_list, 2503)

points = simulate_race(reindeer_list, 2503)
race_winner = max(points.items(), key=lambda item: item[1])[0]
max_points = points[race_winner]

print(f'Part 1: {winner} won with a distance of {distance} km')
print(f'Part 2: {race_winner} won with {max_points} points')

File: 2015/Day14/main.lua
local function calculate_distance(speed, fly_time, rest_time, total_time)
    local cycle_time = fly_time + rest_time
    local full_cycles = math.floor(total_time / cycle_time)
    local remaining_time = total_time % cycle_time

    local distance = full_cycles * fly_time * speed
    if remaining_time > fly_time then
        distance = distance + fly_time * speed
    else
        distance = distance + remaining_time * speed
    end

    return distance
end

local function find_winning_reindeer(reindeer_list, total_time)
    local max_distance = 0
    local winner = nil

    for _, reindeer in ipairs(reindeer_list) do
        local distance = calculate_distance(reindeer.speed, reindeer.fly_time, reindeer.rest_time, total_time)
        if distance > max_distance then
            max_distance = distance
            winner = reindeer.name
        end
    end

    return winner, max_distance
end

local function parse_reindeer(line)
    local pattern = "(%w+) can fly (%d+) km/s for (%d+) seconds, but then must rest for (%d+) seconds."
    local name, speed, fly_time, rest_time = line:match(pattern)
    if name and speed and fly_time and rest_time then
        return {
            name = name,
            speed = tonumber(speed),
            fly_time = tonumber(fly_time),
            rest_time = tonumber(rest_time)
        }
    end
    return nil
end

local function simulate_race(reindeer_list, duration)
    local points = {}
    local distances = {}

    for _, reindeer in ipairs(reindeer_list) do
        points[reindeer.name] = 0
        distances[reindeer.name] = 0
    end

    for second = 1, duration do
        for _, reindeer in ipairs(reindeer_list) do
            if second % (reindeer.fly_time + reindeer.rest_time) <= reindeer.fly_time and second % (reindeer.fly_time + reindeer.rest_time) ~= 0 then
                distances[reindeer.name] = distances[reindeer.name] + reindeer.speed
            end
        end

        local max_distance = 0
        for _, distance in pairs(distances) do
            if distance > max_distance then
                max_distance = distance
            end
        end

        for _, reindeer in ipairs(reindeer_list) do
            if distances[reindeer.name] == max_distance then
                points[reindeer.name] = points[reindeer.name] + 1
            end
        end
    end

    return points
end

local reindeer_list = {}
for line in io.lines("input.txt") do
    local reindeer = parse_reindeer(line)
    if reindeer then
        table.insert(reindeer_list, reindeer)
    end
end

local winner, distance = find_winning_reindeer(reindeer_list, 2503)
local points = simulate_race(reindeer_list, 2503)

local race_winner = nil
local max_points = 0
for name, point in pairs(points) do
    if point > max_points then
        max_points = point
        race_winner = name
    end
end

print(string.format("Part 1: %s won with %d km", winner, distance))
print(string.format("Part 2: %s won with %d points", race_winner, max_points))
